i.MX EPIT Timer Driver

stanislav-poboril opened the pull request:
This adds EPIT driver and enables it on M4 core of i.MX 6SoloX SoC and on UDOO Neo Full board.
The driver conforms to the new counter API (#8340, #8331) and should be merged into topic-counters branch.

stanislav-poboril committed:
imx: Add IMX EPIT driver for i.MX socs

Add shim driver for i.MX EPIT (Enhanced Periodic Interrupt Timer)
peripheral which can be used for i.MX6SoloX, i.MX7D and other i.MX socs.

Origin: Original

Signed-off-by: Stanislav Poboril <stanislav.poboril@nxp.com>

stanislav-poboril committed:
imx: Add EPIT peripheral support for i.MX6SoloX soc

Add EPIT (Enhanced Periodic Interrupt Timer) peripheral support
for i.MX6SoloX soc.

Origin: Original

Signed-off-by: Stanislav Poboril <stanislav.poboril@nxp.com>

stanislav-poboril committed:
imx: Enabled EPIT in Udoo Neo Full board

Enabled EPIT (Enhanced Periodic Interrupt Timer) in Cortex-M4 core
of Udoo Neo Full board so the counter API could be used there.

Origin: Original

Signed-off-by: Stanislav Poboril <stanislav.poboril@nxp.com>

galak commented on drivers/counter/Kconfig.imx_epit:
```diff
@@ -0,0 +1,48 @@
+
+if TIMER_IMX_EPIT_1
+
+config TIMER_IMX_EPIT_1_PRESCALER
```
Can we put the prescaler in DTS?

stanislav-poboril replied:
I've seen it set in Kconfig [here](https://github.com/zephyrproject-rtos/zephyr/pull/8340/commits/fb5588d31e18e44841d288ddaa39f0886d1986cc) in #8340, so I wanted to make it the same.
If the most of counter/timer drivers use dts, I can change it.

galak replied:
I think we should move the others to dts as well.

stanislav-poboril replied:
Moved the prescaler setting into DTS for this pr.

galak commented on dts/arm/nxp/nxp_imx6sx_m4.dtsi:
```diff
@@ -241,6 +241,30 @@
+			status = "disabled";
+		};
+
+		epit2:epit@420fffff {
```
I think there's a typo in the UM, I'm guessing the address should be 420D_4000 for EPIT2, since it shows up at 020D_4000.

stanislav-poboril replied:
You are right, thanks. Fixed.

galak reviewed (changes requested):
I think EPIT2 is at wrong address in DTS:

```
devices/MCIMX6X/MCIMX6X_M4.h:#define EPIT1_BASE                               (0x420D0000u)
devices/MCIMX6X/MCIMX6X_M4.h:#define EPIT2_BASE                               (0x420D4000u)
```

MaureenHelm commented on drivers/counter/Kconfig.imx_epit:
```diff
@@ -0,0 +1,26 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+config TIMER_IMX_EPIT
```
Rename this to `COUNTER_IMX_EPIT`

stanislav-poboril replied:
Done.

MaureenHelm commented on drivers/counter/timer_imx_epit.c:
```diff
@@ -0,0 +1,206 @@
+/*
```
Rename the file to `counter_imx_epit.c`

stanislav-poboril replied:
Done.

MaureenHelm commented on boards/arm/udoo_neo_full_m4/Kconfig.defconfig:
```diff
@@ -57,4 +57,14 @@ config GPIO_IMX_PORT_7
+	default y
+
+config TIMER_IMX_EPIT_2
+	default n
```
Can remove these two lines

stanislav-poboril replied:
Done.

MaureenHelm reviewed (changes requested):
Can you also add `counter` to the list of supported features in the board yaml and doc, and update `tests/drivers/counter/counter_basic_api` to run on this board?

stanislav-poboril committed:
tests: drivers: counter: Enabled counter tests on UDOO Neo board

Tests i.MX EPIT timer peripheral driver. Since it supports zero number
of channel alarms, some tests had to be adjusted to look at the
number of channel alarms supported first.

Signed-off-by: Stanislav Poboril <stanislav.poboril@nxp.com>

stanislav-poboril commented on drivers/counter/timer_imx_epit.c:
```diff
@@ -0,0 +1,206 @@
+	EPIT_Type *base = get_epit_config(dev)->base;
+	struct imx_epit_data *driver_data = dev->driver_data;
+
+	if ((EPIT_CR_REG(base) & EPIT_CR_EN_MASK) == 0U) {
```
Fixed.

stanislav-poboril commented on drivers/counter/timer_imx_epit.c:
```diff
@@ -0,0 +1,206 @@
+		/* (Re)enable EPIT Output Compare interrupt */
+		EPIT_SetIntCmd(base, true);
+	} else {
+		/* Reset reload value without affecting the actual count */
```
Removed this.

stanislav-poboril commented:
> Can you also add counter to the list of supported features in the board yaml and doc, and update tests/drivers/counter/counter_basic_api to run on this board?

@MaureenHelm @nordic-krch I've added two commits for the tests, one fixes the compilation errors existing on the `topic-counter` branch at the moment and the other adds support for the UDOO Neo board.
Note that the EPIT timer has just one compare register. As this is used to fire the wrap callback, there are zero channels available. I had to modify the test to look at the number of available channels first and skip some tests if needed.

How about the functionality when only one compare register is available? We have some options:
1. The EPIT implementation could stay as it is now - no channel alarms, wrap callback available.
2. Return error if `counter_set_wrap()` is called with non-null callback. Support one channel alarm.
3. After initialisation of the driver, `counter_get_num_of_channels()` would return 1. If `counter_set_wrap()`
 is called with non-null callback, `counter_get_num_of_channels()` would return 0 and it will not be allowed to set a channel alarm. If `counter_set_wrap()` is called with null callback, `counter_get_num_of_channels()` would still return 1 and allow to set one channel alarm. But I think the `counter_get_num_of_channels()` is meant to be consistent and always number the same number of channels, although it is not stated in the API documentation explicitly.

dbkinder reviewed (approved):
+1 for doc changes

nordic-krch commented:
@stanislav-poboril could you rebase?

regarding the approach when only one channel is available. There are 2 scenarios that can be covered with that:
- periodic alarm
- single-shot alarm.

For periodic alarm you need `set_wrap` to indicate that counter must be reset on wrap. For single shot you can use `set_alarm` or alternatively use `set_wrap` and disable counter in the callback to ensure single-shot functionality. Imo, to keep it simple i would stick with `set_wrap` only. The cost is that user must call `counter_disable` to avoid getting periodic callback.

stanislav-poboril commented:
rebased

MaureenHelm commented:
@nordic-krch 	Please revisit

stanislav-poboril commented:
Rebased and updated after the merge of  #11572.
